### DISCUSSION TITLE
Clarification of what zigpy and quirks layers do

### DISCUSSION THREAD

**Discussion by pipiche38** [zigpy/zigpy#752] [Q&A]
I have started to dig to the different conversation and , I'm not clear on what has to be done on the application layer (which is using zigpy) and what not.

The main question is on Bind cluster and Configure Reporting

When I look to #689 it looks like the application must do it, this would them assume that the application will know the device specifs if any (while the zha-quirks have those as well)

I also have a look to that https://github.com/home-assistant/core/pull/44092#issuecomment-744049638 and here it looks that indeed they were some plan to do such work

What are the recommendations ?

**Comment by Adminiuga** [comment]
Per specs, configure_reporting sets when to send attribute updates on changes and bind controls where to send those updates. Check the ZCL specs. 
In most cases app has to do it, unless you design both the app and device and configure the device to send reports unconditionally, case in point Xiaomi sensors

**Comment by pipiche38** [comment]
This why I was referring to the quirk doing it. As either you have a generic quirk doing what it has to do, either it is a specific one and it does the specific or not ( as Xiaomi).

Is there any specs of zigpy and quirk , so it helps to understand what is/are the role of eahc ?


Envoyé de mon iPhone

**Reply by Adminiuga** [reply]
Zigpy handles zcl and zdo communication/layer of Zigbee. 
Quirk makes non ZCL compliant devices to appear as compliant devices, e.g. emulating battery oercent remaining attribute based on the battery voltage or using a vendor specific cluster and protocol (tuya) to emulate an on/off cluster

**Reply by pipiche38** [reply]
see my response , https://github.com/zigpy/zigpy/discussions/752#discussioncomment-839819

**Comment by Hedda** [comment]
Moved this post from https://github.com/zigpy/zigpy/discussions/582

> to introduce myself I'm developping a Zigbee plugin for Domoticz since more than 3 years , so I know pretty much what all of those devices behaves and works. I don't understand @Hedda and you answers as regards to Xiami motion.
> Let stick to the purpose of this discussion which is Zigbee device initialisation end to end ( as @puddly was mentioning )

> @MattWestb my topic is to have a end to end provisioning done. Today I do understand that it is done by " the upper system", which is in your case HA ?

> My point is to understand what zigpy , quirks does in order to architect the plugin for Domotciz in the right manner. I'm also challenging that fact that if this is not quirks that an extra layer is lacking.

I also understand that binding and attribute reporting settings should be stored at a higher-level, that is, in the application (i.e. application; which in Home Assistant core is the ZHA integration component, but in your case will be [your plugin for Domoticz](https://github.com/pipiche38/Domoticz-Zigpy)). 

I believe that Adminiuga suggested to new developers before to start by looking at a much older version of the Home Assistant integration, presumably because the code there was less complexed then than what code for the newer current Home Assistant integration is today.

> you may look at HA versions about 0.80.0, a bit easier to understand

So I think that he meant that zha component code the older release version of Home Assistant should be easier to understand:

https://github.com/home-assistant/core/tags?after=0.80.3

https://github.com/home-assistant/core/releases/tag/0.80.3

https://github.com/home-assistant/core/releases/tag/0.79.3

Assume that after that the zha integration code was refactored which also had unfortunately of making it harder to understand.

PS: Another suggestion might be to consider buying the [Jeedom Zigbee Plugin](https://github.com/zigpy/zigpy/discussions/725) so get the chance to study its implementation code. This might be a good idea since at least the Jeedom developers had the benefit of getting in later with zigpy so might not have as much legacy as HA has. Downside is that even though they licensed it as open source it is not available in a public repository so you need to first install Jeedom and then buy that Jeedom Zigbee Plugin from their website to download is as as part of the installation.

**Comment by pipiche38** [comment]
I'm sorry but I still do not understand the all stacks. 
Ok so zigpy take care of zcl and zdo communication layer, and I'm fully aligned with that.
On the other side, when I do the pairing of a device whatever brand it is, the zigpy + quirks layers do a full device interview to get 👍 - profileId
DeviceId
Cluster In
Cluster Out

That is not a communication stuff, but a real device interview.

So I feel that we are mixing the zigpy initial idea which is an abstract layer of the Radio controlers, and additional stuff which have been implemented on top of this communication layer.

And that is my point, why an hosting application seating on this set of libs (zigpy + quirks) doing especially the full device discovery is not perfoming the bind + configure reporting set which is basically what is not done to get a full working device .

**Reply by puddly** [reply]
One issue with zigpy handling complete device initialization is that there are often times dozens of attributes per cluster, and many clusters per endpoint, and multiple endpoints per device. What attributes should zigpy setup attribute reporting for? What attributes should be ignored? How often should they report? An application using zigpy will probably end up implementing device "capabilities" based on the supported clusters anyways, which would naturally have attribute reporting configuration:

```python
DEVICE_CAPABILITIES = {
    # Quirks make this cluster show up for Aqara devices, for example, that don't actually expose it but still
    # report temperature changes
    zigpy.zcl.clusters.measurement.TemperatureMeasurement: {
        "commands": {
            ...
        },
        "attributes": {
            "measured_value": {
                "reporting": (30, 900, 50),  # (min_interval, max_interval, reportable_change),
                "display_name": "temperature",
                "converter": ...,
            },
        },
    },
}

# When a new device joins
for endpoint in device.non_zdo_endpoints:
    for cluster in device.output_clusters:
        for cls, cluster_config in DEVICE_CAPABILITIES.items():
            if not isinstance(cluster, cls):
                continue

            await cluster.bind()

            for attr_name, (min_interval, max_interval, reportable_change) in cluster_config["attributes"].items():
                await cluster.configure_reporting(attr_name, min_interval, max_interval, reportable_change)
```

On the other hand, many clusters have really only one useful set of attributes that reporting is set up for. Providing defaults within the cluster itself might not be a bad idea for applications that just want to interact with a Zigbee network and don't care about the specifics:

```python
class TemperatureMeasurement(Cluster):
    cluster_id = 0x0402
    name = "Temperature Measurement"
    ep_attribute = "temperature"
    attributes = {
        # Temperature Measurement Information
        0x0000: ("measured_value", t.int16s),
        0x0001: ("min_measured_value", t.int16s),
        0x0002: ("max_measured_value", t.int16s),
        0x0003: ("tolerance", t.uint16_t),
        # 0x0010: ('min_percent_change', UNKNOWN),
        # 0x0011: ('min_absolute_change', UNKNOWN),
    }
    default_reportable_attributes = {
        "measured_value": (30, 900, 50),
    }
    server_commands = {}
    client_commands = {}
```

**Reply by pipiche38** [reply]
You are totally right and this why I was thinking the Quirk somehow capable to address the configure reporting .

Here is my view, Quirks get a lot of insight on the Device specifics, and having the Config Reporting as well would be a solution.

I'm simply making a parallelism to they way the Domoticz plugin (with zigate works) when geting a new device

full device interview

if during the device interview (based on Model Name and Manufacturer Name) it is a known device, then we apply the config specific requirement like ( Binding and Cfg Reporting )
if the device is not known in the plugin DB, then the interview continue (like it is currently done) by zigpy like:
   - Node Descriptor
   - Cluster Attributes
   - in case of IAS, IAS registration
   - binding Eps which needs to have a configure reporting set. ( this is done with a list of predefine Clusters and Attributes with there Minimum, Maximum, Changes and Timeout default definition

In such , you fully take in charge all device fully compliant, and can eventually fine tune them via the quirk layer (which is maybe not a layer but more a stack as I feel it is at several layers ).

In addition, by doing that you really avoid the complexity of some devices like the Xiaomi to try doing a device discovery when they are even not able to answers. As soon as you get Model Name and Manufacturer Name, you can short cut the all process.

**Reply by Adminiuga** [reply]
Quirk is just a translation from a non ZCL compliant/partially compliant device to a zcl compliant device, so the app has to support only zcl compliant device -- zha integration approach. Another approach like Z2M where the app layer just know how to talk to each particular device. In this case you don't need any quirks

**Comment by pipiche38** [comment]
In addition, you were referring to the Jeedom implementation, from what I see in there code , Jeedom is re-implementing part of the device interview.

So from my side, we might also consider to split zigpy in 2 layers.

zigpy-com which is indeed the initial purpose to be an abstract layer , and then I would agree that it will be up to the host to do what ever it has to do ( device interview, bind, reporting ..... )
zigpy-device-enrollement which is responsible to do the all Application Device provisioning (including the bind + pairing )

**Reply by Hedda** [reply]
If zigpy intended purpose as an abstraction layer is meant to present a full standard Zigbee stack then might it not be logical to have ZCL (Zigbee Cluster Library) and ZDO (Zigbee Device Object) commands, and in the case of ZDO bind request commands for ZDO Bind Request commands should part of ZDO set of commands, but it is the application which utilizes ZDO bind request commands to only send the command and then the bindings / binding table are stored locally on each device?

https://github.com/MarkDing/IoT-Developer-Boot-Camp/wiki/Introduction-of-Zigbee-Basic

https://www.silabs.com/community/wireless/zigbee-and-thread/knowledge-base.entry.html/2017/12/28/creating_bindingsan-egzv

**Reply by pipiche38** [reply]
@Hedda fair, so it doesn't make sense for me to spend more time on Zigpy for Domoticz ... keep what you are doing for HA and enjoy
@puddly thanks for your constructive approach and making the zigpy lib much powerfull

I have see in 1 year some good evolution, like for instance the device signature has really improve and prevent the application to interview the device , may be in one year, zigpy will do a full device onboarding

**Reply by Hedda** [reply]
> @Hedda fair, so it doesn't make sense for me to spend more time on Zigpy for Domoticz ... keep what you are doing for HA and enjoy

OK, understandable. Sad thought that you chose to hide or delete the repo and not just leave it as archived for prosperity.
 
https://github.com/pipiche38/Domoticz-Zigpy -> 404

**Reply by pipiche38** [reply]
Does that mean that you would be considering working on the Domoticz integration and not having a dogmatic view on zigpy.

As already mentioned from an Architecture standpoint this is no sense as currently zigpy is doing incomplete device enrolment. you can say that Quirks is to make non-standard device seen as standard, but just imagine how the above layer will have to handle such non-standard device presented as standard and start setting bind and configure reporting.

Just try to make the exercise on a Tuya TS0601 type

If that is sad for you, fair, please accept my apologize, but doesn't make sense for me to leave uncompleted , not-working and not Documented work for others.

**Reply by Adminiuga** [reply]
Zigpy doesn't do any device enrollment. It just discovers enough information so it could talk to the device. The rest is up to the level above it. You want zigpy to decide how and what attributes to configure and that is dumb, inflexible and not scalable.
